### ISSUE TITLE
More compact XML presentation for function calls

### ISSUE THREAD

**Issue by napernik** [Orckestra/C1-CMS-Foundation#584]
Currently we're creating an <f:param> xml tag for every function parameter, using xml attributes instead should both make it more readable and reduce the file sizes/improve CRUD performance for XmlDataProvider.

Before:

```
<f:function name="Orckestra.Web.Html.TeaserCards.IconTeaser">
	<f:param name="Icon" value="facebook-square" />
	<f:param name="IconColor" value="primary" />
	<f:param name="Title" value="Facebook" />
	<f:param name="Description" value="Find us on Facebook - contact us or like our page to keep up to date with news." />
	<f:param name="ReadMoreUrl" value="http://www.facebook.com/C1CMS" />
</f:function>
```

After

```
<f:function name="Orckestra.Web.Html.TeaserCards.IconTeaser"
	_Icon="facebook-square"
	_IconColor="primary"
	_Title="Facebook"
	_Description="Find us on Facebook - contact us or like our page to keep up to date with news."
	_ReadMoreUrl="http://www.facebook.com/C1CMS" />
```

**Comment by burningice2866** [comment]
I fail so see how that is more readable. And most editors/formatters will change that into a single (un)readable line which is hardly an improvement?

**Comment by napernik** [comment]
You're right, the example isn't representative. I took a random line from IPagePlaceholderContent data type, that's how it is saved:

```<f:function name="Orckestra.Web.Html.ImageAndTextTeaser" xmlns:f="http://www.composite.net/ns/function/1.0"><f:param name="Image" value="MediaArchive:bb4b8072-3e31-47d6-ada9-84077a86eee3" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Text" xmlns:f="http://www.composite.net/ns/function/1.0"><html><head></head><body><h2>Full Width Image And Text Teaser <br /></h2><p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus hic deserunt iste numquam laborum praesentium doloremque officiis.</p></body></html></f:param><f:param name="BackgroundStyle" value="bg-dark" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Options" value="Is the block full-width" xmlns:f="http://www.composite.net/ns/function/1.0" /></f:function><f:function name="Orckestra.Web.Html.ImageAndTextTeaser" xmlns:f="http://www.composite.net/ns/function/1.0"><f:param name="Image" value="MediaArchive:c50bbc07-8aba-4a53-b73f-7cee6d06a8e9" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Text" xmlns:f="http://www.composite.net/ns/function/1.0"><html><head></head><body><h2>Full Width Image And Text Teaser<br /></h2><p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus hic deserunt iste numquam laborum praesentium doloremque officiis.</p></body></html></f:param><f:param name="BackgroundStyle" value="bg-light" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Options" value="Is the image on the right,Is the block full-width" xmlns:f="http://www.composite.net/ns/function/1.0" /></f:function><f:function name="Orckestra.Web.Html.ImageAndTextTeaser" xmlns:f="http://www.composite.net/ns/function/1.0"><f:param name="Image" value="MediaArchive:79715853-2d99-4a85-918b-3b18d0d48701" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Text" xmlns:f="http://www.composite.net/ns/function/1.0"><html><head></head><body><h1>Image and Text Teaser</h1><p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus hic deserunt iste numquam laborum praesentium doloremque officiis.</p></body></html></f:param><f:param name="BackgroundStyle" value="bg-dark" xmlns:f="http://www.composite.net/ns/function/1.0" /></f:function><f:function name="Orckestra.Web.Html.ImageAndTextTeaser" xmlns:f="http://www.composite.net/ns/function/1.0"><f:param name="Image" value="MediaArchive:8ab4b82a-cb2a-4ef7-954b-f88f52e6783e" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Text" xmlns:f="http://www.composite.net/ns/function/1.0"><html><head></head><body><h1>Image and Text Teaser</h1><p class="lead">Lorem ipsum dolor sit amet, consectetur adipisicing elit.&#160;</p></body></html></f:param><f:param name="BackgroundStyle" value="bg-light" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Options" value="Is the image on the right,Has spacing at the top,Has spacing at the bottom" xmlns:f="http://www.composite.net/ns/function/1.0" /></f:function><f:function name="Orckestra.Web.Html.ImageAndTextTeaser" xmlns:f="http://www.composite.net/ns/function/1.0"><f:param name="Image" value="MediaArchive:91f98350-2475-4daf-be1d-593072a28b6b" xmlns:f="http://www.composite.net/ns/function/1.0" /><f:param name="Text" xmlns:f="http://www.composite.net/ns/function/1.0"><html><head></head><body><h1>Image and Text Teaser</h1><p class="lead">Lorem ipsum dolor sit amet, consectetur adipisicing elit.&#160;</p></body></html></f:param><f:param name="BackgroundStyle" value="bg-primary" xmlns:f="http://www.composite.net/ns/function/1.0" /></f:function>```

Almost completely unreadable without throwing the code into Notepad++/another editor to get some highlighting, as well as xmlns:f="http://www.composite.net/ns/function/1.0" declarations all over the place.

So my thinking is - the reason it is unreadable is a) lack of formatting b) lots of redundant xml declarations that do not mean anything, which could not even be there, should the format be different to begin with, if we were to replace:

```<f:param name="BackgroundStyle" value="bg-light" xmlns:f="http://www.composite.net/ns/function/1.0" />```

With

```_BackgroundStyle="bg-light"```

It wouldn't solve the formatting issue (as well as encoding xml with xml to begin with #268), but should help with the redundancy part.

**Comment by burningice2866** [comment]
My concern is when you write documentation, examples, pass function-call markup around between developers and write it directly in your templates and functions via ie. VS(code) or another favorite editor, that the markup would then be auto-formatted to a single-line-element.

**Comment by aeont** [comment]
Dima,
With _BackgroundStyle="bg-light" you create even more weird markup. I think BackgroundStyle="bg-light" is something in between which could work. But. How <f:param name="Text" xmlns:f="http://www.composite.net/ns/function/1.0"><html><head></head><body><h2>Full Width Image And Text Teaser<br /></h2><p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus hic deserunt iste numquam laborum praesentium doloremque officiis.</p></body></html></f:param> will look like? I assume  <Text><html><head></head>.... </Text>...  and last - what I really don't like... with new markup - there is no way fix old markup to new format with replace manually.. unless write some converter script... or you will write procedure which will detect and convert old format to new one before update.. in this case all good... simple update will convert all.

**Comment by napernik** [comment]
Fixing existing markup into a new format isn't a necessity/big problem, as the parsing can be backwards compatible. Making some upgrade code that can run through existing **IPagePlaceholderContent** also should be relatively quick task.

The other problem with the attribute approach is that there's often a "Name" parameter. So it either have to me changes to something like **_name**, or all attributes have to have a prefix/suffix.

```
<f:function name="Orckestra.Web.Html.TeaserCards.IconTeaser">
	<Icon value="facebook-square" />
	<IconColor value="primary" />
	<Title value="Facebook" />
	<Description value="Find us on Facebook - contact us or like our page to keep up to date with news." />
	<ReadMoreUrl value="http://www.facebook.com/C1CMS" />
</f:function>
```

>  I assume ```<Text><html><head></head>.... </Text>```

Yes

**Comment by burningice2866** [comment]
Parameter name as element name is def. a way to go, but doesn't it cause some issues regarding schema validation and namespaces?
